frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -2444,252 +2489,292 @@ ApplicabilityResult instantiateSignature(Context* context,
-                       context,
-                       resolved,
-                       sig,
-                       resolvedPoiInfo.poiFnIdsUsed(),
-                       resolvedPoiInfo.recursiveFnsUsed());
+    return nullptr;
+  }
 
-  } else {
-    CHPL_ASSERT(false && "this query should be called on Functions");
+  const T& store(ResolutionContext* rc, T x, const Args& args) {
+    static_assert(std::tuple_size_v<Args> == 2);
+    auto sig = std::get<0>(args);
+
+    if (auto rv = rc->findParentResolverFor(sig)) {
+      auto& m = rv->sigAndInfoToChildPtr;
+      ResolvedFunction::SigAndInfo k = { sig, std::get<1>(args) };
+      const ResolvedFunction* v = x;
+      auto check = m.insert(std::make_pair(std::move(k), v));
+      CHPL_ASSERT(check.second);
+      return check.first->second;
+    }
+
+    CHPL_ASSERT(false && "Should not reach here!");
+    auto& f = x != nullptr ? rc->lastFrameOrBaseMutable() : rc->baseFrame_;
+    return f.cache(std::move(x));
+  }
+};
+
+template <typename T, typename Args> struct
+ResolutionContext::UnstableCache<resolveFunctionByPoisQuery, T, Args> {
+
+  const T* fetchOrNull(ResolutionContext* rc, const Args& args) {
+    static_assert(std::tuple_size_v<Args> == 1);
+    auto& poiTrace = std::get<0>(args);
+    auto sig = std::get<0>(poiTrace);
+
+    if (auto rv = rc->findParentResolverFor(sig)) {
+      auto& m = rv->poiTraceToChild;
+      auto it = m.find(poiTrace);
+      if (it != m.end()) return &it->second;
+    }
+
+    return nullptr;
   }
 
-  // Return the unique result from the query (that might have been saved above)
-  const owned<ResolvedFunction>& resolved =
-    resolveFunctionByPoisQueryWrapper(context, sig, resolvedPoiInfo);
+  const T& store(ResolutionContext* rc, T x, const Args& args) {
+    static_assert(std::tuple_size_v<Args> == 1);
+    auto& poiTrace = std::get<0>(args);
+    auto sig = std::get<0>(poiTrace);
 
-  const ResolvedFunction* result = resolved.get();
+    if (auto rv = rc->findParentResolverFor(sig)) {
+      auto& m = rv->poiTraceToChild;
+      auto p = std::make_pair(poiTrace, std::move(x));
+      auto check = m.insert(std::move(p));
+      CHPL_ASSERT(check.second);
+      return check.first->second;
+    }
 
-  return QUERY_END(result);
+    CHPL_ASSERT(false && "Should not reach here!");
+    auto& f = x != nullptr ? rc->lastFrameOrBaseMutable() : rc->baseFrame_;
+    return f.cache(std::move(x));
+  }
+};
+
+static const owned<ResolvedFunction>&
+resolveFunctionByPoisQuery(ResolutionContext* rc, PoiInfo::Trace poiTrace) {
+  CHPL_RESOLUTION_QUERY_BEGIN(resolveFunctionByPoisQuery, rc, poiTrace);
+
+  auto& rcq = CHPL_RESOLUTION_REF_TO_CURRENT_QUERY_HANDLE();
+  if (rcq.canUseGlobalCache()) {
+    // We should not have made it this far if we can recall from the global
+    // context query cache. The query guard should have returned early.
+    CHPL_ASSERT(false && "Should be set in 'resolveFunctionByInfoQuery'!");
+  }
+
```

Comment:
We lost this comment. Could you add it back in?

```
  // the actual value is set in resolveFunctionByInfoQuery after it is
  // computed because computing it generates the poiFnIdsUsed which is
  // part of the key for this query.
  ```